File: libavogadro/src/python/mesh.cpp
// Last update: timvdm 18 June 2009
#include <boost/python.hpp>

#include <avogadro/mesh.h>
#include <avogadro/color3f.h>
#include <avogadro/primitive.h>
#include <avogadro/molecule.h>

#include <QColor>

using namespace boost::python;
using namespace Avogadro;

bool reserve(Mesh &self, unsigned int size)
{
  return self.reserve(size);
}

void export_Mesh()
{
  
  class_<Avogadro::Mesh, bases<Avogadro::Primitive>, boost::noncopyable>("Mesh", no_init)
    
    //
    // read-only poperties 
    //
    .add_property("valid", 
        &Mesh::valid, 
        "Sanity checking function - is the mesh sane?")

    //
    // read/write properties
    //
    .add_property("name", 
        &Mesh::name, 
        &Mesh::setName, 
        "The name of the Mesh.")

    .add_property("stable", 
        &Mesh::stable, 
        &Mesh::setStable, 
        "Indicate whether the Mesh is complete or currently being modified. In "
        "general using Mesh values from an unstable Mesh is not advisable.")

    .add_property("isoValue", 
        &Mesh::isoValue, 
        &Mesh::setIsoValue, 
        "The iso value used to generate the Mesh.")

    .add_property("otherMesh", 
        &Mesh::otherMesh, 
        &Mesh::setOtherMesh, 
        "The unique id of the other Mesh if this is part of a pair.")

    .add_property("cube", 
        &Mesh::cube, 
        &Mesh::setCube, 
        "The unique id of the Cube the Mesh was generated from.")

    .add_property("vertices", 
        make_function(&Mesh::vertices, return_value_policy<return_by_value>()), 
        &Mesh::setVertices, 
        "List containing all of the vertices in a one dimensional array.")

    .add_property("numVertices", 
        &Mesh::numVertices, 
        "The number of vertices.")

    .add_property("numNormals", 
        &Mesh::numNormals, 
        "The number of normals.")

    .add_property("normals", 
        make_function(&Mesh::normals, return_value_policy<return_by_value>()), 
        &Mesh::setNormals, 
        "List containing all of the normals in a one-dimensional array.")

    .add_property("colors", 
        make_function(&Mesh::colors, return_value_policy<return_by_value>()),
        &Mesh::setColors)
 
    // real functions
    .def("reserve", 
        &Mesh::reserve, 
        "Reserve the expected space for the mesh. This causes all member vector "
        "storage to call the reserve function with the number specified.")
    .def("reserve", 
        reserve, 
        "Reserve the expected space for the mesh. This causes all member vector "
        "storage to call the reserve function with the number specified.")

    .def("vertex", &Mesh::vertex, return_value_policy<return_by_value>()) // FIXME
    .def("normal", &Mesh::normal, return_value_policy<return_by_value>()) // FIXME
    .def("color", &Mesh::color, return_value_policy<return_by_value>()) // FIXME
    
    .def("clear", 
        &Mesh::clear, 
        "Clear all mesh data.")
    
    .def("addVertices", 
        &Mesh::addVertices, 
        "Add one or more vertices, i.e., the vector is expected to be of length "
        "3 x n where n is an integer.")

    .def("addNormals", 
        &Mesh::addNormals,
        "Add one or more normals, i.e., the vector is expected to be of length "
        "3 x n where n is an integer.")

    .def("addColors", 
        &Mesh::addColors,
        "Add one or more colors, i.e., the vector is expected to be of length "
        "3 x n where n is an integer.")
    ;
   
}

File: libavogadro/src/python/unittest/mesh.py
import Avogadro
import unittest
from numpy import *

class TestMesh(unittest.TestCase):
  def setUp(self):
    self.mesh = Avogadro.Mesh()

  def test_name(self):
    self.mesh.name = "testing"
    self.assertEqual(self.mesh.name, "testing")
 
  def test_reserve(self):
    # just check if the method is there
    self.assertEqual(self.mesh.reserve(100, True), True)

  def test_stable(self):
    self.mesh.stable = False
    self.assertEqual(self.mesh.stable, False)
    self.mesh.stable = True
    self.assertEqual(self.mesh.stable, True)

  def test_vertices(self):
    vertices = []
    vertices.append(array([0., 0., 1.]))
    vertices.append(array([1., 0., 0.]))
    vertices.append(array([0., 1., 0.]))

    self.mesh.vertices = vertices
    self.assertEqual(len(self.mesh.vertices), 3)
    self.assertEqual(self.mesh.vertex(0)[2], 1)
    
    self.mesh.addVertices(vertices)
    self.assertEqual(len(self.mesh.vertices), 6)
 
  def test_normals(self):
    normals = []
    normals.append(array([0., 0., 1.]))
    normals.append(array([1., 0., 0.]))
    normals.append(array([0., 1., 0.]))

    self.mesh.normals = normals
    self.assertEqual(len(self.mesh.normals), 3)
    self.assertEqual(self.mesh.normal(0)[2], 1)
 
    self.mesh.addNormals(normals)
    self.assertEqual(len(self.mesh.normals), 6)
  
  def test_valid(self):
    # just test it's there...
    self.assertEqual(self.mesh.valid, True)
 
  def test_normals(self):
    l = []
    l.append(array([0., 0., 1.]))
    l.append(array([1., 0., 0.]))
    l.append(array([0., 1., 0.]))

    self.mesh.vertices = l
    self.mesh.normals = l
    
    self.mesh.clear()
    self.assertEqual(len(self.mesh.vertices), 0)
    self.assertEqual(len(self.mesh.normals), 0)
  

if __name__ == "__main__":
  unittest.main()
